eth/watchers/roundswatcher.go

```diff
@@ -0,0 +1,160 @@
+package watchers
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/livepeer/go-livepeer/eth/contracts"
+
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/livepeer/go-livepeer/eth/blockwatch"
+)
+
+// RoundsWatcher is a type alias for a thread safe in-memory cache that watches on-chain events for new initialized rounds
+type RoundsWatcher struct {
+	mu                   sync.RWMutex
+	lastInitializedRound *big.Int
+	blockHashForRound    [32]byte
+
+	quit chan struct{}
+
+	watcher Watcher
+	lpEth   eth.LivepeerEthClient
+	dec     *EventDecoder
+}
+
+type Watcher interface {
+	Subscribe(sink chan<- []*blockwatch.Event) event.Subscription
+}
+
+// NewRoundsWatcher creates a new instance of RoundsWatcher and sets the initial cache through an RPC call to an ethereum node
+func NewRoundsWatcher(watcher Watcher, lpEth eth.LivepeerEthClient, addr ethcommon.Address) (*RoundsWatcher, error) {
+	dec, err := NewEventDecoder(addr, contracts.RoundsManagerABI)
+	if err != nil {
+		return nil, fmt.Errorf("error creating decoder: %v", err)
+	}
+
+	return &RoundsWatcher{
+		quit:    make(chan struct{}),
+		watcher: watcher,
+		lpEth:   lpEth,
+		dec:     dec,
+	}, nil
+}
+
+// LastInitializedRound gets the last initialized from cache
+func (rw *RoundsWatcher) LastInitializedRound() (*big.Int, error) {
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	return rw.lastInitializedRound, nil
+}
+
+// BlockHashForRound gets the blockhash for the last initialized round from cache
+func (rw *RoundsWatcher) BlockHashForRound(round *big.Int) ([32]byte, error) {
+	if round == nil {
+		return [32]byte{}, fmt.Errorf("round is nil")
+	}
+	lastRound, _ := rw.LastInitializedRound()
+	if lastRound.Cmp(round) != 0 {
+		return rw.lpEth.BlockHashForRound(round)
+	}
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	return rw.blockHashForRound, nil
+}
+
+// setLastInitializedRound sets the last initialized round in cache
+func (rw *RoundsWatcher) setLastInitializedRound(round *big.Int) {
```

Comment:
I think we should use a single setter that locks -> updates the round and block hash -> unlocks. Otherwise, it could be possible for a reader to read an updated round, but stale block hash:

Ex.

Writer updates old round to new round
Reader fetches new round
Reader fetches old block hash
Writer updates old block hash to new block hash

Reply:
fixed in 80d3e25